File: Week_08/cpp/src/_57_InsertInterval.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    // O(n)
    // 我没有想到要将重叠区间合并，也少考虑了一种情况
    // 我一直纠结于修改原来的数组。
    vector<vector<int>> insert(vector<vector<int>> &intervals, vector<int> &newInterval)
    {
        // 往以左端点排序的无重叠区间列表中插入一个新区间
        // 假设新区间为[l, r]，则考虑列表中每一个区间[li, ri]
        // 如果[li, ri]如果[li, ri]有重叠则合并[li, ri]到[l, r]
        // 最后将合并后的[l, r]加入到结果数组中
        vector<vector<int>> ret;
        int l = newInterval[0], r = newInterval[1];
        bool placed = false;
        for (const auto &it : intervals)
        {
            if (it[1] < l)
            {
                ret.push_back(it);
            }
            else if (r < it[0])
            {
                if (!placed)
                {
                    ret.push_back({l, r});
                    placed = true;
                }
                ret.push_back(it);
            }
            else
            {
                l = min(l, it[0]);
                r = max(r, it[1]);
            }
        }
        if (!placed)
        {
            ret.push_back({l, r});
        }
        return ret;
    }
};

File: Week_08/kotlin/src/_57_InsertInterval.kt
class _57_InsertInterval_ {
    class Solution {
        fun insert(intervals: Array<IntArray>, newInterval: IntArray): Array<IntArray> {
            // 因为区间列表是按左端点排序的， 那么与新区间重叠的区间一定是连续的
            // 那么我们可以顺序扫描列表，将与新区间重叠的区间合并到新区间中，与新区间不重叠的区间按原顺序放在列表中
            var l = newInterval[0]
            var r = newInterval[1]
            val ret = arrayListOf<IntArray>()
            var placed = false
            intervals.forEach {
                when {
                    it[1] < l -> ret += it
                    it[0] > r -> {
                        if (!placed) {
                            ret += intArrayOf(l, r)
                            placed = true
                        }
                        ret += it
                    }
                    else -> {
                        l = minOf(l, it[0])
                        r = maxOf(r, it[1])
                    }
                }
            }
            if (!placed) {
                ret += intArrayOf(l, r)
            }
            return ret.toTypedArray()
        }
    }
}